compiler/runtime/OMRRelocationRecord.cpp

```diff
@@ -0,0 +1,559 @@
+         for (int32_t *offsetPtr = offsetsBase;offsetPtr < endOfOffsets; offsetPtr+=2)
+            {
+            int32_t offsetHigh = *offsetPtr;
+            int32_t offsetLow = *(offsetPtr+1);
+            uint8_t *reloLocationHigh = reloOrigin + offsetHigh + 2; // Add 2 to skip the first 16 bits of instruction
+            uint8_t *reloLocationLow = reloOrigin + offsetLow + 2; // Add 2 to skip the first 16 bits of instruction
+            int32_t rc = self()->applyRelocation(reloRuntime, reloTarget, reloLocationHigh, reloLocationLow);
+            if (rc != 0)
+               {
+               return rc;
+               }
+            }
+         }
+      else
+         {
+         int16_t *offsetsBase = (int16_t *) (((uint8_t*)_record) + self()->bytesInHeader(reloTarget));
+         int16_t *endOfOffsets = (int16_t *) self()->nextBinaryRecord(reloTarget);
+         for (int16_t *offsetPtr = offsetsBase;offsetPtr < endOfOffsets; offsetPtr+=2)
+            {
+            int16_t offsetHigh = *offsetPtr;
+            int16_t offsetLow = *(offsetPtr+1);
+            uint8_t *reloLocationHigh = reloOrigin + offsetHigh + 2; // Add 2 to skip the first 16 bits of instruction
+            uint8_t *reloLocationLow = reloOrigin + offsetLow + 2; // Add 2 to skip the first 16 bits of instruction
+            int32_t rc = self()->applyRelocation(reloRuntime, reloTarget, reloLocationHigh, reloLocationLow);
+            if (rc != 0)
+               {
+               return rc;
+               }
+            }
+         }
+      }
+   else
+      {
+      if (self()->wideOffsets(reloTarget))
+         {
+         int32_t *offsetsBase = (int32_t *) (((uint8_t*)_record) + self()->bytesInHeader(reloTarget));
+         int32_t *endOfOffsets = (int32_t *) self()->nextBinaryRecord(reloTarget);
+         for (int32_t *offsetPtr = offsetsBase;offsetPtr < endOfOffsets; offsetPtr++)
+            {
+            int32_t offset = *offsetPtr;
+            uint8_t *reloLocation = reloOrigin + offset;
+            int32_t rc = self()->applyRelocation(reloRuntime, reloTarget, reloLocation);
+            if (rc != 0)
+               {
+               return rc;
+               }
+            }
+         }
+      else
+         {
+         int16_t *offsetsBase = (int16_t *) (((uint8_t*)_record) + self()->bytesInHeader(reloTarget));
+         int16_t *endOfOffsets = (int16_t *) self()->nextBinaryRecord(reloTarget);
+         for (int16_t *offsetPtr = offsetsBase;offsetPtr < endOfOffsets; offsetPtr++)
+            {
+            int16_t offset = *offsetPtr;
+            uint8_t *reloLocation = reloOrigin + offset;
+            int32_t rc = self()->applyRelocation(reloRuntime, reloTarget, reloLocation);
+            if (rc != 0)
+               {
+               return rc;
+               }
+            }
+         }
+      }
+      return 0;
+   }
+   
+
+uint32_t OMR::RelocationRecord::_relocationRecordHeaderSizeTable[TR_NumExternalRelocationKinds] =
```

Comment:
There's a problem with this table. These sizes are based on the various binary templates that are defined in OpenJ9. If you look at the equivalent table in OpenJ9 [1], you'll see that the table is mostly filled with `sizeof`s. Now you can't do that here because then you'd have to move those binary template definitions from OpenJ9 over to OMR. However, that doesn't make sense because those binary templates are very much OpenJ9 specific. In fact, pretty much every RelocationKind right now is OpenJ9 Specific. The unfortunate thing is that the RelocationKind definition sits in OMR.

So, I guess what I'm getting at is, i think this table should be filled with maybe -1 or something. Nothing in OMR uses it or even _could_ use it. In the future, when the work to have OpenJ9 use the OMR AOT infra gets under way, this table can then be rearchitected appropriately.


[1] https://github.com/eclipse/openj9/blob/fd96afe5f58b9125d812dcb45ae85f7376e0b7d3/runtime/compiler/runtime/RelocationRecord.cpp#L4614

Reply:
In an offline chat with @georgkrylov, the eventual way to fix this would be to have static size methods in each relo record. However, for this PR the table will just be filled with 0s.